### PR TITLE
Reentrency

### PR DESCRIPTION
**Pull request by Kerilk**
This fixes most of the reentrancy issues I found in the code (see https://github.com/CHIP-SPV/chip-spv/issues/485).

Each commit is self contained except for c856c36cf9e3b9cbcdcdf18e9515f15a8b3be0fe which requires 8143082a6a8662064a3ca8efa89ce90c894e2f76 . Nonetheless, subsequent commits depend on previous ones due to the depth of the recursions in our implementation of the API.

~There are still some reentrancy left in the graph implementation:~

```
src/CHIPGraph.cc:    auto Status = hipMemcpy(Dst_, Src_, Count_, Kind_);
src/CHIPGraph.cc:    auto Status = hipMemcpy3DAsync(&Params_, Queue);
src/CHIPGraph.cc:  auto Status = hipEventRecord(Event_, Queue);
src/CHIPGraph.cc:  auto Status = hipStreamWaitEvent(Queue, Event_, Flags);
src/CHIPGraph.hh:        hipMemcpyFromSymbol(Dst_, Symbol_, SizeBytes_, Offset_, Kind_);
src/CHIPGraph.hh:    auto Status = hipMemcpyToSymbol(Symbol_, Src_, SizeBytes_, Offset_, Kind_);
```

~But these will have to be handled in a different PR.~

I fixed those as well, I have kept the semantic used in their current implementation even if in some case (memcpy, memcpy to and from symbol) I think the given queue should be used and the node be executed in a non-blocking fashion.

I had to integrate the stream per thread API entry points in the main binding file, and took the opportunity to fix them as they were obviously broken (not returning errors). I think the PER_THREAD_DEFAULT_STREAM option in CMake should be investigated, I don't think it is required (or even functional), and if it is, it reveals an issue, as the stream per thread usage of the library is an application compile time parameter, not a library compile time parameter.

### CODE REVIEW

File: src/CHIPBindings.cc [hunk 1]
```diff
@@ -2323,38 +2466,70 @@ hipError_t hipHostMalloc(void **Ptr, size_t Size, unsigned int Flags) {
...
-  RETURN(hipHostMalloc(Ptr, Size, Flags));
+hipError_t hipMallocHost(void **Ptr, size_t Size) {
+  CHIP_TRY
+  CHIPInitialize();
+  RETURN(hipHostMalloc_internal(Ptr, Size, hipHostMallocDefault));
+  CHIP_CATCH
 }
 
-hipError_t hipFree(void *Ptr) {
+DEPRECATED("use hipHostMalloc instead")
```
Comment (linehill) [inline src/CHIPBindings.cc:2487]:
```suggestion
```
Could be removed. Users won't and see this and developers shouldn't.

**Comment by linehill** [comment]
Hi, could you apply these clang-format suggestions:
```diff
--- src/CHIPBindings.cc	(before formatting)
+++ src/CHIPBindings.cc	(after formatting)
@@ -2132,9 +2132,9 @@
   CHIP_CATCH
 }
 
-static inline hipError_t
-hipStreamWaitEvent_internal(hipStream_t Stream, hipEvent_t Event,
-                            unsigned int Flags) {
+static inline hipError_t hipStreamWaitEvent_internal(hipStream_t Stream,
+                                                     hipEvent_t Event,
+                                                     unsigned int Flags) {
   auto ChipQueue = static_cast<CHIPQueue *>(Stream);
   auto ChipEvent = static_cast<CHIPEvent *>(Event);
 
@@ -3063,7 +3063,8 @@
   NULLCHECK(Dst);
 
   auto ChipQueue = Backend->getActiveDevice()->getDefaultQueue();
-  auto Res = hipMemset2DAsync_internal(Dst, Pitch, Value, Width, Height, ChipQueue);
+  auto Res =
+      hipMemset2DAsync_internal(Dst, Pitch, Value, Width, Height, ChipQueue);
   if (Res == hipSuccess)
     ChipQueue->finish();
 
@@ -3071,9 +3072,9 @@
   CHIP_CATCH
 }
 
-static inline hipError_t
-hipMemset3DAsync_internal(hipPitchedPtr PitchedDevPtr, int Value,
-                          hipExtent Extent, hipStream_t Stream) {
+static inline hipError_t hipMemset3DAsync_internal(hipPitchedPtr PitchedDevPtr,
+                                                   int Value, hipExtent Extent,
+                                                   hipStream_t Stream) {
   auto ChipQueue = Backend->findQueue(static_cast<CHIPQueue *>(Stream));
   const hipMemsetParams Params = {
       /* Dst */ PitchedDevPtr.ptr,
@@ -3124,8 +3125,8 @@
       size_t SizeBytes = Width;
       auto Offset = i * (Pitch * PitchedDevPtr.ysize) + j * Pitch;
       char *DstP = (char *)Dst;
-      auto Res = hipMemsetAsync_internal(DstP + Offset, Value, SizeBytes,
-                                         Stream);
+      auto Res =
+          hipMemsetAsync_internal(DstP + Offset, Value, SizeBytes, Stream);
       if (Res != hipSuccess)
         break;
     }
@@ -3157,7 +3158,8 @@
   CHIP_CATCH
 }
 
-static inline hipError_t hipMemset_internal(void *Dst, int Value, size_t SizeBytes) {
+static inline hipError_t hipMemset_internal(void *Dst, int Value,
+                                            size_t SizeBytes) {
   logInfo("hipMemset(Dst={}, Value={}, SizeBytes={})", Dst, Value, SizeBytes);
 
   char CharVal = Value;
@@ -3206,7 +3208,7 @@
   CHIP_TRY
   CHIPInitialize();
   NULLCHECK(Dst);
-  RETURN(hipMemset_internal(Dst, Value,SizeBytes));
+  RETURN(hipMemset_internal(Dst, Value, SizeBytes));
   CHIP_CATCH
 }
 
@@ -3351,10 +3353,10 @@
 }
 
 static inline hipError_t
-hipMemcpy2DToArrayAsync_internal(hipArray *Dst, size_t WOffset,
-                                 size_t HOffset, const void *Src,
-                                 size_t SPitch, size_t Width, size_t Height,
-                                 hipMemcpyKind Kind, hipStream_t Stream) {
+hipMemcpy2DToArrayAsync_internal(hipArray *Dst, size_t WOffset, size_t HOffset,
+                                 const void *Src, size_t SPitch, size_t Width,
+                                 size_t Height, hipMemcpyKind Kind,
+                                 hipStream_t Stream) {
   auto ChipQueue = Backend->findQueue(static_cast<CHIPQueue *>(Stream));
   const hipMemcpy3DParms Params = {
       /* hipArray_t srcArray */ nullptr,
@@ -3411,9 +3413,8 @@
   NULLCHECK(Dst, Src);
   auto ChipQueue = Backend->getActiveDevice()->getDefaultQueue();
 
-  auto Res = hipMemcpy2DToArrayAsync_internal(Dst, WOffset, HOffset, Src,
-                                              SPitch, Width, Height, Kind,
-                                              ChipQueue);
+  auto Res = hipMemcpy2DToArrayAsync_internal(
+      Dst, WOffset, HOffset, Src, SPitch, Width, Height, Kind, ChipQueue);
 
   if (Res == hipSuccess)
     ChipQueue->finish();
@@ -3425,9 +3426,8 @@
 static inline hipError_t
 hipMemcpy2DFromArrayAsync_internal(void *Dst, size_t DPitch,
                                    hipArray_const_t Src, size_t WOffset,
-                                   size_t HOffset, size_t Width,
-                                   size_t Height, hipMemcpyKind Kind,
-                                   hipStream_t Stream) {
+                                   size_t HOffset, size_t Width, size_t Height,
+                                   hipMemcpyKind Kind, hipStream_t Stream) {
   auto ChipQueue = Backend->findQueue(static_cast<CHIPQueue *>(Stream));
   const hipMemcpy3DParms Params = {
       /* hipArray_t srcArray */ const_cast<hipArray_t>(Src),
@@ -3504,9 +3504,8 @@
   NULLCHECK(Dst, Src);
   auto ChipQueue = Backend->getActiveDevice()->getDefaultQueue();
 
-  auto Res = hipMemcpy2DFromArrayAsync_internal(Dst, DPitch, Src, WOffset,
-                                                HOffset, Width, Height, Kind,
-                                                ChipQueue);
+  auto Res = hipMemcpy2DFromArrayAsync_internal(
+      Dst, DPitch, Src, WOffset, HOffset, Width, Height, Kind, ChipQueue);
   if (Res == hipSuccess)
     ChipQueue->finish();
 
@@ -3574,8 +3573,8 @@
     CHIPERR_LOG_AND_THROW("Copy size greater than allocation size",
                           hipErrorTbd);
 
-  RETURN(hipMemcpy_internal((char *)DstArray->data + DstOffset, SrcHost,
-                            Count, hipMemcpyHostToDevice));
+  RETURN(hipMemcpy_internal((char *)DstArray->data + DstOffset, SrcHost, Count,
+                            hipMemcpyHostToDevice));
 
   CHIP_CATCH
 }
@@ -3778,8 +3777,8 @@
   CHIP_TRY
   CHIPInitialize();
   NULLCHECK(Symbol, Src);
-  RETURN(hipMemcpyToSymbolAsync_internal(Symbol, Src, SizeBytes, Offset,
-                                         Kind, Stream));
+  RETURN(hipMemcpyToSymbolAsync_internal(Symbol, Src, SizeBytes, Offset, Kind,
+                                         Stream));
   CHIP_CATCH
 }
 
@@ -3826,8 +3825,8 @@
   CHIP_TRY
   CHIPInitialize();
   NULLCHECK(Dst, Symbol);
-  RETURN(hipMemcpyFromSymbolAsync_internal(Dst, Symbol, SizeBytes, Offset,
-                                           Kind, Stream));
+  RETURN(hipMemcpyFromSymbolAsync_internal(Dst, Symbol, SizeBytes, Offset, Kind,
+                                           Stream));
   CHIP_CATCH
 }
 
@@ -3885,10 +3884,10 @@
   CHIP_CATCH
 }
 
-static inline hipError_t
-hipLaunchKernel_internal(const void *HostFunction, dim3 GridDim,
-                         dim3 BlockDim, void **Args, size_t SharedMem,
-                         hipStream_t Stream) {
+static inline hipError_t hipLaunchKernel_internal(const void *HostFunction,
+                                                  dim3 GridDim, dim3 BlockDim,
+                                                  void **Args, size_t SharedMem,
+                                                  hipStream_t Stream) {
   auto ChipQueue = Backend->findQueue(static_cast<CHIPQueue *>(Stream));
   if (ChipQueue->captureIntoGraph<CHIPGraphNodeKernel>(
           HostFunction, GridDim, BlockDim, Args, SharedMem)) {
@@ -4104,14 +4103,11 @@
   CHIP_CATCH
 }
 
-static inline hipError_t
-hipModuleLaunchKernel_internal(hipFunction_t Kernel, unsigned int GridDimX,
-                               unsigned int GridDimY, unsigned int GridDimZ,
-                               unsigned int BlockDimX, unsigned int BlockDimY,
-                               unsigned int BlockDimZ,
-                               unsigned int SharedMemBytes,
-                               hipStream_t Stream, void *KernelParams[],
-                               void *Extra[]) {
+static inline hipError_t hipModuleLaunchKernel_internal(
+    hipFunction_t Kernel, unsigned int GridDimX, unsigned int GridDimY,
+    unsigned int GridDimZ, unsigned int BlockDimX, unsigned int BlockDimY,
+    unsigned int BlockDimZ, unsigned int SharedMemBytes, hipStream_t Stream,
+    void *KernelParams[], void *Extra[]) {
   auto ChipQueue = Backend->findQueue(static_cast<CHIPQueue *>(Stream));
 
   if (KernelParams == Extra)
@@ -4172,10 +4168,9 @@
                                  void *Extra[]) {
   CHIP_TRY
   CHIPInitialize();
-  RETURN(hipModuleLaunchKernel_internal(Kernel, GridDimX, GridDimY, GridDimZ,
-                                        BlockDimX, BlockDimY, BlockDimZ,
-                                        SharedMemBytes, Stream, KernelParams,
-                                        Extra));
+  RETURN(hipModuleLaunchKernel_internal(
+      Kernel, GridDimX, GridDimY, GridDimZ, BlockDimX, BlockDimY, BlockDimZ,
+      SharedMemBytes, Stream, KernelParams, Extra));
   CHIP_CATCH
 }
 
@@ -4213,11 +4208,10 @@
   if (Result != hipSuccess)
     RETURN(Result);
 
-  Result = hipModuleLaunchKernel_internal(Kernel, GridBlocksX, GridBlocksY,
-                                          GridBlocksZ, LocalWorkSizeX,
-                                          LocalWorkSizeY, LocalWorkSizeZ,
-                                          SharedMemBytes, Stream, KernelParams,
-                                          Extra);
+  Result = hipModuleLaunchKernel_internal(
+      Kernel, GridBlocksX, GridBlocksY, GridBlocksZ, LocalWorkSizeX,
+      LocalWorkSizeY, LocalWorkSizeZ, SharedMemBytes, Stream, KernelParams,
+      Extra);
   if (Result != hipSuccess)
     RETURN(Result);
 
@@ -4542,10 +4536,10 @@
                          hipMemcpyKind kind) {
   CHIP_TRY
   CHIPInitialize();
-  hipError_t Res = hipMemcpyAsync_internal(dst, src, sizeBytes, kind,
-                                           hipStreamPerThread);
+  hipError_t Res =
+      hipMemcpyAsync_internal(dst, src, sizeBytes, kind, hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4556,11 +4550,10 @@
   CHIP_TRY
   CHIPInitialize();
   NULLCHECK(symbol, src);
-  hipError_t Res = hipMemcpyToSymbolAsync_internal(symbol, src, sizeBytes,
-                                                   offset, kind,
-                                                   hipStreamPerThread);
+  hipError_t Res = hipMemcpyToSymbolAsync_internal(
+      symbol, src, sizeBytes, offset, kind, hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4571,11 +4564,10 @@
   CHIP_TRY
   CHIPInitialize();
   NULLCHECK(dst, symbol);
-  hipError_t Res = hipMemcpyFromSymbolAsync_internal(dst, symbol, sizeBytes,
-                                                     offset, kind,
-                                                     hipStreamPerThread);
+  hipError_t Res = hipMemcpyFromSymbolAsync_internal(
+      dst, symbol, sizeBytes, offset, kind, hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4588,7 +4580,7 @@
   hipError_t Res = hipMemcpy2DAsync_internal(dst, dpitch, src, spitch, width,
                                              height, kind, hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4599,11 +4591,11 @@
   CHIP_TRY
   CHIPInitialize();
   NULLCHECK(dst, src);
-  hipError_t Res = hipMemcpy2DToArrayAsync_internal(dst, wOffset, hOffset, src,
-                                                    spitch, width, height,
-                                                    kind, hipStreamPerThread);
+  hipError_t Res =
+      hipMemcpy2DToArrayAsync_internal(dst, wOffset, hOffset, src, spitch,
+                                       width, height, kind, hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4619,7 +4611,7 @@
                                                       hOffset, width, height,
                                                       kind, hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4630,7 +4622,7 @@
   NULLCHECK(p);
   hipError_t Res = hipMemcpy3DAsync_internal(p, hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4639,10 +4631,10 @@
   CHIP_TRY
   CHIPInitialize();
   NULLCHECK(dst);
-  hipError_t Res = hipMemsetAsync_internal(dst, value, sizeBytes,
-                                           hipStreamPerThread);
+  hipError_t Res =
+      hipMemsetAsync_internal(dst, value, sizeBytes, hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4655,7 +4647,7 @@
   hipError_t Res = hipMemset2DAsync_internal(dst, pitch, value, width, height,
                                              hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
@@ -4668,7 +4660,7 @@
   hipError_t Res = hipMemset3DAsync_internal(pitchedDevPtr, value, extent,
                                              hipStreamPerThread);
   if (hipSuccess != Res)
-      RETURN(Res);
+    RETURN(Res);
   RETURN(hipStreamSynchronize_internal(hipStreamPerThread));
   CHIP_CATCH
 }
```

File: src/CHIPBindings.cc [hunk 2]
```diff
@@ -1319,10 +1320,97 @@ hipError_t hipMemcpyPeerAsync(void *Dst, int DstDeviceId, const void *Src,
   CHIP_CATCH
 };
 
-hipError_t hipMemcpyParam2DAsync(const hip_Memcpy2D *PCopy,
-                                 hipStream_t Stream) {
+static inline hipError_t hipMemcpyAsync_internal(void *Dst, const void *Src,
```
Comment (linehill) [inline src/CHIPBindings.cc:1323]: Should we use a naming convention closer to the LLVM style?
```suggestion
static inline hipError_t hipMemcpyAsyncInternal(void *Dst, const void *Src,
```

File: src/CHIPBindings.cc (hunk 1 above)
Comment (Kerilk) [reply src/CHIPBindings.cc]: Agreed, and addressed in https://github.com/CHIP-SPV/chip-spv/pull/486/commits/4adf00e8cf3e4069b556b302226cfc4e9a8a6431

**Comment by Kerilk** [comment]
@pvelesko  I may need a bit of help with the CI here.
I ran the unit tests on our gen9 implementation (L0) on JLSE, and the only tests that failed are:

	529 - Unit_hipHostMalloc_Basic (Failed)
	713 - ABM_AddKernel_MultiTypeMultiSize - int (Failed)
	714 - ABM_AddKernel_MultiTypeMultiSize - long (Failed)
	715 - ABM_AddKernel_MultiTypeMultiSize - float (Failed)
	716 - ABM_AddKernel_MultiTypeMultiSize - long long (Failed)
	717 - ABM_AddKernel_MultiTypeMultiSize - double (Failed)

Requested total local size 512 exceeds HW limit 256
or
Requested total local size 500 exceeds HW limit 256

But I can't seem to reproduce the other issues. Any advice here?

**Comment by Kerilk** [comment]
I seem to be able to reproduce 2 of the errors on the JLSE gen9 OpenCL backend (plus the same errors from the L0 backend):

```
	480 - Unit_hipMemset2D_UniqueWidthHeight (Failed)
	560 - Unit_hipMemsetFunctional_ZeroSize_2D (Failed)
```
So I should be able to make progress.

**Comment by pvelesko** [comment]
> I think the PER_THREAD_DEFAULT_STREAM option in CMake should be investigated, I don't think it is required (or even functional), and if it is, it reveals an issue, as the stream per thread usage of the library is an application compile time parameter, not a library compile time parameter.

You are correct on that this shouldn't be a cmake macro. We can get rid of this, but I don't think we support setting per-thread as default using `hipcc`

**Comment by pvelesko** [comment]
> @pvelesko I may need a bit of help with the CI here. I ran the unit tests on our gen9 implementation (L0) on JLSE, and the only tests that failed are:
> 
> ```
> 529 - Unit_hipHostMalloc_Basic (Failed)
> 713 - ABM_AddKernel_MultiTypeMultiSize - int (Failed)
> 714 - ABM_AddKernel_MultiTypeMultiSize - long (Failed)
> 715 - ABM_AddKernel_MultiTypeMultiSize - float (Failed)
> 716 - ABM_AddKernel_MultiTypeMultiSize - long long (Failed)
> 717 - ABM_AddKernel_MultiTypeMultiSize - double (Failed)
> ```
> 
> Requested total local size 512 exceeds HW limit 256 or Requested total local size 500 exceeds HW limit 256
> 
> But I can't seem to reproduce the other issues. Any advice here?
> 
> Edit: I pushed a small fix but that shouldn't impact the CI results.

1. Exceeding hw limits should not happen anywhere - I've fixed the tests that were causing this. If you are seeing these errors on gen9 we should investigate
2. The CI errors you saw earlier were due to the new driver - I left a message on slack warning about this. I have since rolled back some changes to the CI so I will be re-running & merging this branch today.

**Comment by pvelesko** [comment]
@Kerilk  I looked at the errors - the ABM kernels are supposed to fail due to HW limits. I thought that both CI iGPU JLSE iGPU had the same hw limits which is not the case. 

`Unit_hipHostMalloc_Basic` - needs to be investigated.

**Comment by Kerilk** [comment]
For hipHostMalloc it gave me a:
```
Requested total local size 512 exceeds HW limit 256
```
I'll add the full log for you in a bit

**Pull request merged**